src/packs/ignored.rs

```diff
@@ -0,0 +1,89 @@
+use std::collections::HashSet;
+
+pub fn is_ignored(rules: &HashSet<String>, path: &str) -> anyhow::Result<bool> {
+    let (allow_list, deny_list): (HashSet<&String>, HashSet<&String>) =
+        rules.iter().partition(|rule| rule.starts_with('!'));
+
+    // allow-list (starts with !) takes precedence over deny-list (does not start with !
+    if allow_list.iter().any(|rule| is_match(&rule[1..], path)) {
+        return Ok(false);
+    }
+    if deny_list.iter().any(|rule| is_match(rule, path)) {
+        return Ok(true);
+    }
+
+    Ok(false)
+}
+
+fn is_match(rule: &str, path: &str) -> bool {
+    match fnmatch_regex2::glob_to_regex(rule) {
+        Ok(regex) => regex.is_match(path),
+        _ => false,
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[macro_export]
+    macro_rules! test_ignore {
```

Comment:
Dang, you're fancy with these macros.

I haven't gotten into the rust macros yet. I've never liked meta-programming in Ruby, but mostly because it's hard to verify. I love how Rust's meta-programming is statically type checked!